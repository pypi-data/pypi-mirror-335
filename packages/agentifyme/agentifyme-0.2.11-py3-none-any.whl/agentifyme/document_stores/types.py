from typing import Any

from pydantic import BaseModel, Field
from pydantic.json_schema import SkipJsonSchema


class Document(BaseModel):
    """Class representing a document.
    """

    id: str = Field()
    content: str = Field(description="The content of the document.")
    metadata: dict[str, Any] = Field(description="The metadata associated with the document.")


class VectorDocument(Document):
    """Represents a document with an embedding."""

    embedding: list[float] = SkipJsonSchema()
