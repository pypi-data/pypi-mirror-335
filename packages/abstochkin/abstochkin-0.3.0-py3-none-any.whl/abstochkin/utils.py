"""
Some utility functions on generating random number streams, measuring a
function's runtime, calculating a statistic measuring the goodness of fit
when comparing time series data, and performing unit conversion of
kinetic parameters.
"""

#  Copyright (c) 2024-2025, Alex Plakantonakis.
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import functools
from contextlib import contextmanager
from time import perf_counter

import numpy as np
from numpy import random
from scipy.constants import N_A

from .logging_config import logger

logger = logger.getChild(os.path.basename(__file__))


def rng_streams(n: int, random_state: int):
    """
    Generate independent streams of random numbers spawned from the
    same initial seed.

    Parameters
    ----------
    n : int
        number of generators/streams to generate.
    random_state : int, optional
        initial seed from which n new seeds are spawned.

    Returns
    -------
    list[PCG64DXSM Generator objects]
        List of `n` generators of independent streams of random numbers

    Notes
    -----
    See https://numpy.org/doc/stable/reference/random/parallel.html for more info.

    On PCG64DXSM:
        - https://numpy.org/doc/stable/reference/random/upgrading-pcg64.html#upgrading-pcg64
        - https://numpy.org/doc/stable/reference/random/bit_generators/pcg64dxsm.html

    Examples
    --------
    >>> rng = rng_streams(5)  # make 5 random number generators
    >>> a = rng[1].integers(1, 10, 100)
    """
    ss = random.SeedSequence(random_state)
    seeds = ss.spawn(n)

    # return [random.default_rng(seed) for seed in seeds]  # PCG64 generator objects

    # Return PCG64DXSM generator objects
    return [random.Generator(random.PCG64DXSM(seed)) for seed in seeds]


def measure_runtime(fcn):
    """ Decorator for measuring the duration of a function's execution. """
    @functools.wraps(fcn)
    def inner(*args, **kwargs):
        """ This is a wrapper function that measures the execution time of the
        decorated function (`fcn`) and logs it. """
        start = perf_counter()
        fcn(*args, **kwargs)
        stop = perf_counter()
        duration = stop - start
        if duration < 60:
            msg = f"Simulation Runtime: {duration:.3f} sec"
        elif 60 <= duration < 3600:
            msg = f"Simulation Runtime: {duration / 60:.3f} min"
        else:
            msg = f"Simulation Runtime: {duration / 3600:.3f} hr"

        logger.info(msg)

    return inner


def r_squared(actual: np.array, theoretical: np.array) -> float:
    """
    Compute the coefficient of determination, $R^2$.

    In the case of comparing the average AbStochKin-simulated species
    trajectory to its deterministic trajectory. Since the latter is only
    meaningful for a homogeneous population, R² should be
    close to `1` for a simulated homogeneous process.
    For a heterogeneous process, it can be interpreted as how close
    the simulated trajectory is to the deterministic trajectory of a
    *homogeneous* process. In this case, $R^2$ would not be expected
    to be close to $1$ and the importance of looking at this metric
    is questionable.

    Parameters
    ----------
    actual : numpy.array
        Actual data obtained through a simulation.
    theoretical : numpy.array
        Theoretical data to compare the actual data to.

    Returns
    -------
    float
        The coefficient of determination, $R^2$.
    """
    # sst: total sum of squares for simulation avg trajectory
    sst = np.nansum((actual - np.nanmean(actual)) ** 2)
    # ssr: sum of square residuals (data vs deterministic prediction)
    ssr = np.nansum((actual - theoretical) ** 2)

    if sst == 0:
        logger.warning("The total sum of squares is zero. R^2 is undefined.")
        return np.nan
    else:
        return 1 - ssr / sst


def macro_to_micro(macro_val: float | int | list[float | int, ...] | tuple[float | int, float | int],
                   volume: float | int,
                   order: int = 0,
                   *,
                   inverse: bool = False) -> float | list[float, ...] | tuple[float, float]:
    """
    Convert a kinetic parameter value from macroscopic to microscopic form.

    The ABK algorithm uses microscopic kinetic constants, thus necessitating
    the conversion of any molar quantities to their microscopic counterpart.
    For a kinetic parameter, the microscopic form is interpreted as the number
    of transition events per second (or whatever the time unit may be).
    For a molar quantity, its microscopic form is the number of particles in
    the given volume.

    Parameters
    ----------
    macro_val : float or int
        The value of the parameter to be converted, expressed in terms of
        molar quantities.
    volume : float or int
        The volume, in liters, in which the process that the given parameter
        value is a descriptor of.
    order : int, default: 0
        The order of the process whose kinetic parameter is to be converted.
        The default value of 0 is for parameters (such as Km or K50) whose
        units are molarity.
    inverse : bool, default: False
        Perform the inverse of this operation. That is, convert
        from microscopic to macroscopic form.

    Returns
    --------
    float
        A kinetic parameter is returned in units of reciprocal seconds.
        A molar quantity is returned as the number of particles in the
        given volume.

    Notes
    -----
    - A kinetic parameter for a 1st order process will remain unchanged
    because its units are already reciprocal seconds.

    Reference
    ---------
    Plakantonakis, Alex. “Agent-based Kinetics: A Nonspatial Stochastic Method
    for Simulating the Dynamics of Heterogeneous Populations.”
    OSF Preprints, 26 July 2019. Web. Section 2.1.
    (Updated version of the paper in "/docs/Agent-basedKinetics_monograph.pdf")
    """
    try:
        assert volume > 0, "The volume has to be a positive quantity."
        assert order >= 0, "The process order cannot be negative."
    except AssertionError:
        logger.error(f"Assertion(s) failed: {volume=}, {order=}\n"
                     f"The volume has to be a positive quantity. "
                     f"The process order cannot be negative.")
        raise

    denom = (N_A * volume) ** (order - 1) if not inverse else 1 / (N_A * volume) ** (order - 1)

    if isinstance(macro_val, (list, tuple)):
        try:
            assert all([True if val >= 0 else False for val in macro_val]), \
                "The parameter values cannot be negative."
            micro_vals = [val / denom for val in macro_val]
            return micro_vals if isinstance(macro_val, list) else tuple(micro_vals)
        except AssertionError:
            logger.error(f"Assertion failed: The parameter values cannot be negative. "
                         f"{macro_val=}")
            raise
    else:
        try:
            assert macro_val >= 0, "The parameter value cannot be negative."
            return macro_val / denom
        except AssertionError:
            logger.error(f"Assertion failed: The parameter value cannot be negative. "
                         f"{macro_val=}")
            raise


@contextmanager
def log_exceptions():
    """
    A context manager for logging exceptions.

    Use this context manager to log any exceptions that occur within
    the given code block. The exception is re-raised after logging.

    Example
    -------
    with log_exceptions():
        # Code that may raise an exception
    """
    try:
        yield
    except Exception as e:
        logger.error("Exception occurred.", exc_info=e)
        raise
