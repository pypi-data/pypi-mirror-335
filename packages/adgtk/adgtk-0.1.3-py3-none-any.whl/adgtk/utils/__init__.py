"""Common Utils"""

from .cli import get_user_input, get_more_ask, create_line, clear_screen
from .file import load_data_from_csv_file
from .formatting import (
    get_timestamp_now,
    process_possible_yaml,
    count_prefix_spaces)
from .logs import start_logging
from .plotting import plot_single_line
from .processing import (
    string_to_bool,
    get_pair_from_iterable,
    get_sample_from_iterable)
from .settings import load_settings
from .text import camel_case_generation
