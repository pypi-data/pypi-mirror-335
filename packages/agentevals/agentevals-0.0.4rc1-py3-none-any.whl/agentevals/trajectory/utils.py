__all__ = [
    "_is_trajectory_superset",
    "_extract_tool_calls",
    "_get_matcher_for_tool_name",
]

import json

from agentevals.types import (
    ChatCompletionMessage,
    ToolArgsMatchMode,
    ToolArgsMatchOverrides,
)
from typing import Callable, Optional


def _normalize_tool_call(tool_call: dict) -> dict:
    if "function" in tool_call:
        return {
            "name": tool_call["function"]["name"],
            "args": json.loads(tool_call["function"]["arguments"]),
        }
    else:
        return tool_call


def _extract_tool_calls(messages: list[ChatCompletionMessage]) -> list[dict]:
    tool_calls: list[dict] = []
    for message in messages:
        if "tool_calls" in message:
            tool_calls.extend(
                _normalize_tool_call(tool_call) for tool_call in message["tool_calls"]
            )
    return tool_calls


def _is_trajectory_superset(
    outputs: list[ChatCompletionMessage],
    reference_outputs: list[ChatCompletionMessage],
    tool_args_match_mode: ToolArgsMatchMode,
    tool_args_match_overrides: Optional[ToolArgsMatchOverrides] = None,
):
    output_tool_calls = _extract_tool_calls(outputs)
    reference_tool_calls = _extract_tool_calls(reference_outputs)

    # Keep track of which reference tool calls have been matched
    matched_reference_calls = set()

    # For each reference tool call, find a matching output tool call
    for ref_call in reference_tool_calls:
        ref_name = ref_call["name"]
        ref_args = ref_call["args"]

        found_match = False
        for out_idx, out_call in enumerate(output_tool_calls):
            out_name = out_call["name"]

            # Names must match
            if ref_name != out_name:
                continue

            # If we're already using this output call for a different match, skip
            if out_idx in matched_reference_calls:
                continue

            # Check tool args according to match mode
            matcher = _get_matcher_for_tool_name(
                ref_name, tool_args_match_mode, tool_args_match_overrides
            )

            out_args = out_call["args"]
            if matcher(out_args, ref_args):
                matched_reference_calls.add(out_idx)
                found_match = True
                break

        # If we didn't find a match for this reference call, we're not a superset
        if not found_match:
            return False

    return True


def _exact_match(tool_call: dict, reference_tool_call: dict) -> bool:
    return tool_call == reference_tool_call


def _none_match(tool_call: dict, reference_tool_call: dict) -> bool:
    return True


def _get_matcher_for_comparison_mode(
    mode: ToolArgsMatchMode,
) -> Callable[[str, str], bool]:
    if mode == "exact":
        return _exact_match
    else:
        return _none_match


def _get_partial_matcher_on_keys(keys: list[str]) -> Callable[[str, str], bool]:
    def get_nested_value(d: dict, key_path: str):
        current = d
        for part in key_path.split("."):
            current = (
                current["function"]["arguments"]
                if part == "function.arguments"
                else current[part]
            )
        return current

    def matcher(output_call: dict, reference_call: dict) -> bool:
        return all(
            get_nested_value(output_call, key) == get_nested_value(reference_call, key)
            for key in keys
        )

    return matcher


def _get_matcher_for_tool_name(
    tool_call_name: str,
    tool_args_match_mode: ToolArgsMatchMode,
    tool_args_match_overrides: Optional[ToolArgsMatchOverrides],
) -> Callable[[str, str], bool]:
    matcher = _get_matcher_for_comparison_mode(tool_args_match_mode)
    if tool_args_match_overrides is not None and tool_args_match_overrides.get(
        tool_call_name, None
    ):
        override = tool_args_match_overrides.get(tool_call_name)
        if isinstance(override, str):
            matcher = _get_matcher_for_comparison_mode(override)
        elif isinstance(override, Callable):
            matcher = override
        else:
            matcher = _get_partial_matcher_on_keys(override)
    return matcher
