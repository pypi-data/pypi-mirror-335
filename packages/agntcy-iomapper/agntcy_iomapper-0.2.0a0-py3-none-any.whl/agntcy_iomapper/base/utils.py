import copy
import json
import logging
from typing import Any, Dict, List, Optional, Type

import jsonref

logger = logging.getLogger(__name__)


def create_type_from_schema(
    json_schema: Dict[str, Any], json_paths: List[str]
) -> Optional[Type]:
    """
    Creates a new model with only the specified fields from a JSON schema.

    Args:
        json_schema: The JSON schema of the original object.
        json_paths: A list of field names to include in the new model.

    Returns:
        A new Pydantic model class containing only the specified fields.
    """

    # replace $refs with actual object definition
    flatten_json = jsonref.loads(json.dumps(json_schema))

    properties = flatten_json.get("properties", {})

    filtered_properties = {}
    curr_path_schema = {}

    for path in json_paths:
        parts = path.split(".")
        curr_key = parts[0]
        curr_path_schema.clear()

        if curr_key in properties:
            # perform a deepcopy to keep original properties intact
            curr_object_def = copy.deepcopy(properties.get(curr_key))

            curr_path_schema[curr_key] = curr_object_def
            if "anyOf" in curr_object_def:
                sub_schemas = curr_object_def.get("anyOf")

                for i, sub_schema in enumerate(sub_schemas):
                    if "properties" in sub_schema:
                        _props = sub_schema.get("properties", {})
                        curr_path_schema[curr_key]["anyOf"][i]["properties"] = (
                            _get_properties(1, parts, _props, flatten_json)
                        )
                    elif "items" in sub_schema:
                        _curr_items = curr_object_def.get("items", {})
                        curr_path_schema[curr_key]["anyOf"][i]["items"] = (
                            _get_properties(1, parts, _curr_items, flatten_json)
                        )

            if "properties" in curr_object_def:
                props = curr_object_def.get("properties", {})
                curr_path_schema[curr_key]["properties"] = _get_properties(
                    1, parts, props, flatten_json
                )
            elif "items" in curr_object_def:
                curr_items = curr_object_def.get("items", {})
                curr_path_schema[curr_key]["items"] = _get_properties(
                    1, parts, curr_items, flatten_json
                )

        else:
            curr_path_schema[curr_key] = {"type": "object", "properties": {}}
            if len(parts) == 1:
                curr_path_schema[curr_key]["properties"] = flatten_json
            else:
                curr_path_schema[curr_key]["properties"] = _get_properties(
                    1, parts, {}, flatten_json
                )
        if curr_key not in filtered_properties:
            filtered_properties[curr_key] = copy.deepcopy(curr_path_schema[curr_key])
        else:
            filtered_properties[curr_key] = _merge_paths(
                curr_path_schema, filtered_properties
            )[curr_key]

    return filtered_properties


def _merge_paths(d1: dict, d2: dict):
    """
    Merges two dictionaries recursively, combining values for common keys,
    and merging nested dictionaries at the same level.

    Args:
        d1 (dict): The first dictionary.
        d2 (dict): The second dictionary.

    Returns:
        dict: The merged dictionary.
    """
    merged = {}
    keys = set(d1.keys()) | set(d2.keys())

    for key in keys:
        if key in d1 and key in d2:
            if isinstance(d1[key], dict) and isinstance(d2[key], dict):
                merged[key] = _merge_paths(d1[key], d2[key])
            elif isinstance(d1[key], list) and isinstance(d2[key], list):
                merged[key] = []
                for val1, val2 in zip(d1[key], d2[key]):
                    if isinstance(val1, dict) and isinstance(val2, dict):
                        merged[key].append(_merge_paths(val1, val2))
                    else:
                        merged[key].append(val2)

                merged[key].extend(d1[key][len(merged[key]) :])
                merged[key].extend(d2[key][len(merged[key]) :])

            else:
                merged[key] = d2[key]
        elif key in d1:
            merged[key] = d1[key]
        else:
            merged[key] = d2[key]

    return merged


def _get_properties(level, parts, props, json_schema):
    if level >= len(parts):
        return props

    curr_key = parts[level]
    if level >= len(parts) - 1 and len(props) == 0:
        return {curr_key: json_schema}

    if curr_key in props:
        curr_obj = props.get(curr_key)

        if "properties" in curr_obj:
            curr_properties = curr_obj.get("properties", {})
            curr_obj["properties"] = _get_properties(
                level + 1, parts, curr_properties, json_schema
            )
        elif "items" in curr_obj:
            curr_properties = curr_obj.get("items", {})
            curr_obj["items"] = _get_properties(
                level + 1, parts, curr_properties, json_schema
            )
        elif "anyOf" in curr_obj:
            sub_schemas = curr_obj.get("anyOf")
            anyOfs = []
            for sub_schema in sub_schemas:
                s = _get_properties(level + 2, parts, sub_schema, json_schema)
                anyOfs.append(s)
            curr_obj["anyOf"] = anyOfs

        return {curr_key: curr_obj}

    elif "anyOf" in props:
        curr_obj = {curr_key: {"anyOf": []}}
        anyOf = []
        sub_schemas = props.get("anyOf")

        for sub_schema in sub_schemas:
            anyOf.append(_get_properties(level + 1, parts, sub_schema, json_schema))

        curr_obj[curr_key]["anyOf"] = anyOf
        return {curr_key: curr_obj}

    else:
        curr_obj = {curr_key: {"type": "object", "properties": {}}}
        curr_obj[curr_key]["properties"] = _get_properties(
            level + 1, parts, {}, json_schema
        )
        return curr_obj


def extract_nested_fields(data: Any, fields: List[str]) -> dict:
    """Extracts specified fields from a potentially nested data structure
    Args:
        data: The input data (can be any type)
        fields: A list of fields path (e.g.. "fielda.fieldb")
    Returns:
        A dictionary containing the extracted fields and their values.
        Returns empty dictionary if there are errors
    """
    if not fields:
        return {}

    results = {}

    for field_path in fields:
        try:
            value = _get_nested_value(data, field_path)
            results[field_path] = value
        except (KeyError, TypeError, AttributeError, ValueError) as e:
            logger.error(f"Error extracting field {field_path}: {e}")
    return results


def _get_nested_value(data: Any, field_path: str) -> Optional[Any]:
    """
    Recursively retrieves a value from a nested data structure
    """
    current = data
    parts = field_path.split(".")

    for part in parts:
        if isinstance(current, dict):
            current = current[part]
        elif isinstance(current, list) and part.isdigit():
            current = current[int(part)]
        elif hasattr(current, part):
            current = getattr(current, part)
        else:
            current = None

    return current
