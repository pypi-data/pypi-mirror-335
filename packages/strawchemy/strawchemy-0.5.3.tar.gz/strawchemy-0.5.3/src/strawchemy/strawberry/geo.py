from __future__ import annotations

import json
from typing import Any, NewType

from geoalchemy2 import WKBElement, WKTElement
from geoalchemy2.shape import to_shape
from geojson_pydantic.geometries import Geometry as PydanticGeometry
from shapely import Geometry, to_geojson

import strawberry
from pydantic import TypeAdapter

__all__ = ("GeoJSON", "StrawberryGeoComparison")

_PydanticGeometryType = TypeAdapter(PydanticGeometry)


def _serialize_geojson(val: Geometry | WKTElement | WKBElement) -> dict[str, Any]:
    if isinstance(val, WKBElement | WKTElement):
        val = to_shape(val)
    return json.loads(to_geojson(val))


def _parse_geojson(val: dict[str, Any]) -> dict[str, Any]:
    _PydanticGeometryType.validate_python(val)
    return val


GeoJSON = strawberry.scalar(
    NewType("GeoJSON", object),
    description=(
        "The `GeoJSON` type represents GEOJson values as specified by "
        "[RFC 7946](https://datatracker.ietf.org/doc/html/rfc7946)"
    ),
    serialize=_serialize_geojson,
    parse_value=_parse_geojson,
    specified_by_url="https://datatracker.ietf.org/doc/html/rfc7946",
)


class StrawberryGeoComparison:
    contains_geometry: GeoJSON | None = strawberry.UNSET
    within_geometry: GeoJSON | None = strawberry.UNSET
    is_null: bool | None = strawberry.UNSET
