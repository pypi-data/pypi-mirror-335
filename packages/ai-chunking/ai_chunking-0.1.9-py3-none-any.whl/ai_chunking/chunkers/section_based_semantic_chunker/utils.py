from difflib import SequenceMatcher
import re
from pathlib import Path
from typing import Dict, List, Tuple
import asyncio
import tiktoken
from litellm import AsyncOpenAI

from ai_chunking.utils.json_utils import parse_json_response
from ai_chunking.chunkers.section_based_semantic_chunker.prompts import SYSTEM_PROMPT


MODEL_NAME = "gpt-4o-mini"


async def process_with_llm(prompt: str, model_name: str = MODEL_NAME) -> Tuple[str, Dict]:
    """Process text with LLM and return summary and metadata"""
    client = AsyncOpenAI()
    response = await client.chat.completions.create(
        model=model_name,
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": prompt}
        ],
        response_format={ "type": "json_object" }  # Ensure JSON response
    )
    
    # Parse structured response
    response_json = parse_json_response(response.choices[0].message.content)
    return response_json


async def run_concurrent_tasks(tasks: List[asyncio.Task], max_concurrent: int = 5):
    """Run tasks concurrently with semaphore"""
    semaphore = asyncio.Semaphore(max_concurrent)
    
    async def wrapped_task(task):
        async with semaphore:
            return await task
    
    return await asyncio.gather(
        *(wrapped_task(task) for task in tasks),
        return_exceptions=True
    )



