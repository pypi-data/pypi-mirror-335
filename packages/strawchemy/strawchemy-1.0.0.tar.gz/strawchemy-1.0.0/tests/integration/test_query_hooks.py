from __future__ import annotations

from typing import TYPE_CHECKING

import pytest
from strawchemy import StrawchemyAsyncRepository, StrawchemySyncRepository

import strawberry
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

from .types import FilteredFruitType, FruitTypeWithDescription, strawchemy
from .typing import RawRecordData

if TYPE_CHECKING:
    from syrupy.assertion import SnapshotAssertion

    from .fixtures import QueryTracker

pytestmark = [pytest.mark.integration]


@strawberry.type
class AsyncQuery:
    fruits: list[FruitTypeWithDescription] = strawchemy.field(repository_type=StrawchemyAsyncRepository)
    fruits_paginated: list[FruitTypeWithDescription] = strawchemy.field(
        repository_type=StrawchemyAsyncRepository, pagination=True
    )
    user_fruits: list[FilteredFruitType] = strawchemy.field(repository_type=StrawchemyAsyncRepository)


@strawberry.type
class SyncQuery:
    fruits: list[FruitTypeWithDescription] = strawchemy.field(repository_type=StrawchemySyncRepository)
    fruits_paginated: list[FruitTypeWithDescription] = strawchemy.field(
        repository_type=StrawchemySyncRepository, pagination=True
    )
    user_fruits: list[FilteredFruitType] = strawchemy.field(repository_type=StrawchemySyncRepository)


@pytest.fixture
def sync_query() -> type[SyncQuery]:
    return SyncQuery


@pytest.fixture
def async_query() -> type[AsyncQuery]:
    return AsyncQuery


@pytest.mark.parametrize("fruits_query", ["fruits", "fruitsPaginated"])
@pytest.mark.snapshot
async def test_load_columns_hook(
    fruits_query: str,
    any_query: AnyQueryExecutor,
    raw_fruits: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    result = await maybe_async(any_query(f"{{ {fruits_query} {{ description }} }}"))

    assert not result.errors
    assert result.data
    assert result.data[fruits_query] == [
        {"description": f"The {raw_fruits[0]['name']} is {', '.join(raw_fruits[0]['adjectives'])}"},
        {"description": f"The {raw_fruits[1]['name']} is {', '.join(raw_fruits[1]['adjectives'])}"},
        {"description": f"The {raw_fruits[2]['name']} is {', '.join(raw_fruits[2]['adjectives'])}"},
        {"description": f"The {raw_fruits[3]['name']} is {', '.join(raw_fruits[3]['adjectives'])}"},
        {"description": f"The {raw_fruits[4]['name']} is {', '.join(raw_fruits[4]['adjectives'])}"},
    ]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize("fruits_query", ["fruits", "fruitsPaginated"])
async def test_empty_query_hook(fruits_query: str, any_query: AnyQueryExecutor) -> None:
    result = await maybe_async(any_query(f"{{ {fruits_query} {{ emptyQueryHook }} }}"))

    assert not result.errors
    assert result.data
    assert len(result.data[fruits_query]) == 5
    assert result.data[fruits_query] == [{"emptyQueryHook": "success"} for _ in range(5)]


@pytest.mark.snapshot
async def test_custom_query_hook(
    any_query: AnyQueryExecutor, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    result = await maybe_async(any_query("{ userFruits { name } }"))

    assert not result.errors
    assert result.data
    assert len(result.data["userFruits"]) == 1
    assert result.data["userFruits"] == [{"name": "Apple"}]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
