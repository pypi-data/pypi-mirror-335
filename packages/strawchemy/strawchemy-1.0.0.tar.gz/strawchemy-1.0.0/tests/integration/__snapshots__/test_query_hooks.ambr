# serializer version: 1
# name: test_always_load_column_hook[session-tracked-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruits-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruits-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruits-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruitsPaginated-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruitsPaginated-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_always_load_column_hook[session-tracked-fruitsPaginated-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_always_load_column_hook[session-tracked-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_custom_query_hook[session-tracked-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_custom_query_hook[session-tracked-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.id
  FROM fruit AS fruit
  WHERE fruit.name = :name_1
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruits-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruits-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruits-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM fruit AS fruit
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruitsPaginated-async-asyncpg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruitsPaginated-async-psycopg_async_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
# name: test_load_columns_hook[session-tracked-fruitsPaginated-sync-psycopg_engine]
  '''
  SELECT fruit.name,
         fruit.adjectives,
         fruit.id
  FROM
    (SELECT fruit.id AS id,
            fruit.name AS name,
            fruit.adjectives AS adjectives
     FROM fruit AS fruit
     LIMIT :param_1
     OFFSET :param_2) AS fruit
  LIMIT :param_3
  '''
# ---
